src/modules/cpu/kernel/transpose.hpp

```diff
@@ -0,0 +1,443 @@
+        Rpp32u *begin = roi;
+        Rpp32u *length = &roi[nDim];
+        Rpp32u *perm = permTensor;
+
+        bool copyInput = true;
+        for(int i = 0; i < nDim; i++)
+            copyInput *= (perm[i] == i);
+
+        // do memcpy of input to output since output order is same as input order
+        if(copyInput)
+        {
+            memcpy(dstPtrTemp, srcPtrTemp, (size_t)(srcGenericDescPtr->strides[0] * sizeof(Rpp32f)));
+        }
+        else
+        {
+            for(int i = 1; i < nDim; i++)
+                srcPtrTemp += begin[i - 1] * srcGenericDescPtr->strides[i];
+
+            if (nDim == 2 && perm[0] == 1 && perm[1] == 0)
+            {
+                // Optimized AVX version for 2D inputs
+                compute_2d_transpose(srcPtrTemp, dstPtrTemp, length[0], length[1], srcGenericDescPtr->strides[1], dstGenericDescPtr->strides[1]);
+            }
+            else if (nDim == 3)
+            {
+                // Optimized AVX version for 3D inputs of shape(x, y, 16) and permutation order (2, 0, 1)
+                if(perm[0] == 2 && perm[1] == 0 && perm[2] == 1 && length[2] == 16)
+                {
+                    Rpp32u height = length[0];
+                    Rpp32u width = length[1];
+                    Rpp32u channels = 16;
+                    Rpp32u bufferLength = width * channels;
+                    Rpp32u alignedLength = (bufferLength / 64) * 64;
+                    Rpp32u vectorIncrement = 64;
+                    Rpp32u vectorIncrementPerChannel = 4;
+
+                    // initialize pointers for 16 channel
+                    Rpp32f *dstPtrChannel[16];
+                    for(int i = 0; i < 16; i++)
+                        dstPtrChannel[i] = dstPtrTemp + i * dstGenericDescPtr->strides[1];
+
+                    // loop over rows
+                    for(int i = 0; i < height; i++)
+                    {
+                        Rpp32f *srcPtrRow = srcPtrTemp;
+
+                        // update temporary pointers for 16 channel
+                        Rpp32f *dstPtrTempChannel[16];
+                        for(int k = 0; k < 16; k++)
+                            dstPtrTempChannel[k] = dstPtrChannel[k];
+
+                        Rpp32u vectorLoopCount = 0;
+                        for( ; vectorLoopCount < alignedLength; vectorLoopCount += vectorIncrement)
+                        {
+                            __m256 pSrc[8];
+                            // load 64 values for source
+                            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow, &pSrc[0]);
+                            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow + 16, &pSrc[1]);
+                            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow + 32, &pSrc[2]);
+                            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow + 48, &pSrc[3]);
+                            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow + 8, &pSrc[4]);
```

Comment:
Are these actually faster than doing a generic for f32 too?

Reply:
Yes